### DISCUSSION TITLE
Using builtins in TGSL

### DISCUSSION THREAD

**Discussion by iwoplaza** [software-mansion/TypeGPU#419] [General]
Using builtins in typed WGSL functions is simple, as we do not need to provide any mechanism to access them in code. For TGSL functions, this becomes trickier. Right now, to use builtins, users can call them directly in the function body, and it will be ✨ **automagically** ✨ added to the argument list of the function. There are a few pitfalls and complexities that come with it though:

1. Devs could think they can use builtins in regular functions, but builtin arguments can only be used in entry functions.
```ts
const foo = tgpu.fn([], f32).implement(() => {
  // No TypeScript error...
  return 0.5 + builtin.position.x;
});
```
2. Hoisting builtin uses to the argument list adds complexity to the resolution algorithm.
3. When using structs to return multiple values from the vertex shader, and into the fragment shader, the field marked as `@builtin(position)` exists and acts like a builtin input into the fragment shader, which means when a dev uses `builtin.position` in a fragment shader, we have to dedup that use.

### Proposed solution

```ts
// Only entry functions can accept builtins as parameter types.
const vertexMain = tgpu
  .vertexFn([builtin.position])
  .implement((position) => {
    //        ^? vec4f
  });
```

**Comment by mhawryluk** [comment]
I like this approach, seems intuitive 👍
